<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">

    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/pygment_trac.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/print.css" media="print">

    <title>mistartools by grenaud</title>
  </head>

  <body>

    <header>
      <div class="container">
        <h1>mistartools</h1>
        <h2> A set of programs for the management of allele count information on a per individual or population basis</h2>

        <section id="downloads">
          <a href="https://github.com/grenaud/mistartools/zipball/master" class="btn">Download as .zip</a>
          <a href="https://github.com/grenaud/mistartools/tarball/master" class="btn">Download as .tar.gz</a>
          <a href="https://github.com/grenaud/mistartools" class="btn btn-github"><span class="icon"></span>View on GitHub</a>
        </section>
      </div>
    </header>

    <div class="container">
      <section id="main_content">
        <h3>
<a id="welcome-to-github-pages" class="anchor" href="#welcome-to-github-pages" aria-hidden="true"><span class="octicon octicon-link"></span></a>Motivation</h3>

<p>Analysis of nuclear genomes often involves combining genotyping information from various individuals or populations to compute summary statistics or exporting them to various file formats (EIGENSTRAT,PLINK, etc..). This is achieved by parsing of multiple genotyping files (ex: VCF) to combined sets which are later exported to various formats or are used to compute summary statistics. </p>

<p>Research groups usually do this by coding custom scripts tailored to the task at hand. New group members often cannot reproduce the analyses that were performed and often code their own scripts from scratch. This leads to time being wasted and an overall lack of reproducibility.</p>

<p>mistartools are a set or programs coded in C++ designed to extract allele counts from VCF files (or 23andme or the raw base count from BAM files) and store it as a compressed representations of allele counts: </p>

<pre><code>
#chr	coord	REF,ALT	IndividualA
7	35190	G,N	2,0:0
</code></pre>

<p>The line above that "IndividualA" had two reference alleles (G) and since no alternative alleles were found, the alternative is an N. The first column is the chromosome name, the second the coordinate, the third the bases for the reference and alternative. We do not consider tri-allelic sites. The fourth column is the allele count for the individual in the following format :

<pre><code>
REF,ALT:CPG
</code></pre>

<p> Where REF is the reference allele count and the ALT is the alternative allele count. The CPG field is a flag for CpG (0 is no CpG, 1 otherwise). In the presence of an alternative allele, the following representations is used:</p>

<pre><code>
#chr	coord	REF,ALT	IndividualA
7	517798	T,C	1,1:0
</code></pre>

<p>The line above says that IndividualA had one T and one C at that position hence a heterozygous site. </p>


<p>mistartools allows users to transform genotyping data into this intermediate allele count. Data from multiple sources can be combined from multiple sources, transformed and exported to numerous formats</p>

<h3>
<a id="designer-templates" class="anchor" href="#designer-templates" aria-hidden="true"><span class="octicon octicon-link"></span></a>File format</h3>

<p>The mistartools file can be divided into two parts, the header that details the operations that were performed and the allele count per se. Here is an example:</p>

<pre><code>
#MISTAR
#PG:/pathtomistartools/vcf2mistar --allowall /pathtovcf/IndividualA.hg19_1000g.4.mod.vcf.gz IndividualA chr4.epo.gz 
#GITVERSION: 8ac3cbfd59c78e74395c8accca9d174a0241ebd5
#DATE: 2013-06-03
#chr	coord	REF,ALT	root	anc	IndividualA
4	72045	A,N	0,0:0	0,0:0	2,0:0
4	72046	T,N	0,0:0	0,0:0	2,0:0
4	72047	T,N	0,0:0	0,0:0	2,0:0
4	72048	T,G	0,0:0	0,0:0	1,1:0
4	72049	G,N	0,0:0	0,0:0	2,0:0
4	72050	T,N	0,0:0	0,0:0	2,0:0
4	72051	A,N	0,0:0	0,0:0	2,0:0
4	72052	G,N	0,0:0	0,0:0	2,0:0
</code></pre>

<p>The header lines started with a # sign. Each operation made on the files from creation to transformation leaves a "watermark" that allows users to know precisely which operation was done and which version of mistartools was used. The remaining lines contain the allele count for chromosome 4 for 8 sites. Three samples are present, the chimp allele, the chimp-human ancestor and an individual named "IndividualA". If multiple indivuals are combined, the allele count lines will appear as:</p>

<pre><code>
#chr	coord	REF,ALT	root	anc	IndividualA	IndividualB	IndividualC
4	72045	A,N	0,0:0	0,0:0	2,0:0	2,0:0	2,0:0
4	72046	T,N	0,0:0	0,0:0	2,0:0	2,0:0	2,0:0
4	72047	T,N	0,0:0	0,0:0	2,0:0	2,0:0	2,0:0
4	72048	T,G	0,0:0	0,0:0	1,1:0	1,1:0	2,0:0
4	72049	G,N	0,0:0	0,0:0	2,0:0	2,0:0	2,0:0
4	72050	T,N	0,0:0	0,0:0	2,0:0	2,0:0	2,0:0
4	72051	A,N	0,0:0	0,0:0	2,0:0	2,0:0	2,0:0
4	72052	G,N	0,0:0	0,0:0	2,0:0	2,0:0	2,0:0
</code></pre>

<p>
We recommend using bgzip to compress the files and index and query them using tabix from <a href="https://github.com/samtools" class="user-mention">htslib</a>.
</p>

<h3>Input data</h3>

<p>
Data can be transformed into mistartools internal format from 
<ul>
<li> VCF
<li> BAM
<li> 23andme
<li> AXT alignments
</ul>
</p>




<h3>Operations on data</h3>

<p>
Using the data in the internal format, here are examples of operations that can be performed:
<ul>
<li> unions or intersections
<li> meld two populations as one
<li> force two populations to share one or all alleles
<li> remove populations
<li> add/remove populations
</ul>

For instance, for obtaining sites where all individuals in specific population are fixed derived the following workflow can be used:
<ol>
<li> transform individual VCF files into mistartools files using "vcf2mistar"
<li> unite each individual files using "mistarintersection" 
<li> meld all individuals into a single population using "mistarmeld"
<li> disable sharing of alleles with the ancestor using "mistarfilter znosharing"
</ol>

</p>

<h3>Exporting data</h3>
<p>
Data can be exported to the following formats:

<ul>
<li> treemix
<li> PLINK
<li> EIGENSTRAT
<li> NEXUS
<li> fasta
</ul>
</p>

<h3>Documentation</h3>

<p>
Please refer to the documentation manual available <a href="https://github.com/grenaud/mistartools/blob/master/doc/reference.pdf?raw=true">here</a>.
</p>
<BR>
<BR>
<BR>


<h3>Questions ? Feature requests ?</h3>

<p>
Contact <a href="https://github.com/grenaud/" class="user-mention">Gabriel Renaud</a>: <BR><BR>

<img src="http://grenaud.github.io/myaddr.png"></img>	 
</p>


    
  </body>
</html>
